### PR TITLE
Nix 2.24.1; nixVersions.latest: 2.23.3 -> 2.24.1

### PR DESCRIPTION
**Pull request by Mic92**
## Description of changes
Changelog: https://github.com/NixOS/nix/blob/2.24.1/doc/manual/src/release-notes/rl-2.24.md

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/tools/package-management/nix/default.nix [hunk 1]
```diff
@@ -184,6 +184,12 @@ in lib.makeExtensible (self: ({
     self_attribute_name = "nix_2_23";
   };
 
+  nix_2_24 = common {
+    version = "2.24.1";
+    hash = "sha256-3yFEvUDPB7GlCMI9I5VV+HXMVOT38h3lnw01nIXU2F4=";
+    self_attribute_name = "nix_2_24";
+  };
```
Comment (emilazy) [inline pkgs/tools/package-management/nix/default.nix:191]: You need the SDK stuff from `nixVersions.git` for this to work on `x86_64-darwin`.

**Review by emilazy — DISMISSED** [review]

File: pkgs/tools/package-management/nix/default.nix (hunk 1 above)
Comment (Mic92) [reply pkgs/tools/package-management/nix/default.nix]: Thanks. Added.

**Comment by JohnRTitor** [comment]
Mic I looked for a existing PR but found none before opening #331788, hopefully you don't mind me meddling with the PR title

**Comment by philiptaron** [comment]
@oxalica `nil` fails to build due to `nix __dump-builtins` returning an error now. It was removed in https://github.com/NixOS/nix/pull/8330.

**Comment by Aleksanaa** [comment]
(Still the same question) @Ericson2314 @tomberek Should we try to build with meson in this release?

**Comment by tomberek** [comment]
No.. I think we do the existing build this time and work on removing it in favor of meson for the next bump. There are still outstanding issues/prs: https://github.com/NixOS/nix/pull/11224

Likely can get the meson build for next NixOS release? Hoping...

**Comment by nyabinary** [comment]
Should this also include a git bump?

**Comment by Aleksanaa** [comment]
Not necessary, we usually bump this separately

**Review by philiptaron — APPROVED** [review]
Result of `nixpkgs-review pr 331720` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>nil</li>
  </ul>
</details>
<details>
  <summary>5 packages built:</summary>
  <ul>
    <li>nixVersions.latest (nixVersions.nix_2_24)</li>
    <li>nixVersions.latest.debug (nixVersions.nix_2_24.debug)</li>
    <li>nixVersions.latest.dev (nixVersions.nix_2_24.dev)</li>
    <li>nixVersions.latest.doc (nixVersions.nix_2_24.doc)</li>
    <li>nixVersions.latest.man (nixVersions.nix_2_24.man)</li>
  </ul>
</details>

`nil` was previously pinged; this PR could update that dependency to pin 2.23 instead of using `nixVersions.latest`.

**Review by llakala — APPROVED** [review]

**Review by philiptaron — APPROVED** [review]
Yeah, that does it. [Thanks @JohnRTitor](https://en.wikipedia.org/wiki/John_Titor).

**Comment by JohnRTitor** [comment]
Result of `nixpkgs-review pr 331720` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>10 packages built:</summary>
  <ul>
    <li>nixVersions.latest</li>
    <li>nixVersions.latest.debug</li>
    <li>nixVersions.latest.dev</li>
    <li>nixVersions.latest.doc</li>
    <li>nixVersions.latest.man</li>
    <li>nixVersions.nix_2_23</li>
    <li>nixVersions.nix_2_23.debug</li>
    <li>nixVersions.nix_2_23.dev</li>
    <li>nixVersions.nix_2_23.doc</li>
    <li>nixVersions.nix_2_23.man</li>
  </ul>
</details>

**Pull request merged**